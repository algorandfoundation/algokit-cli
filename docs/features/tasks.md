# AlgoKit Tasks

AlgoKit Tasks are a collection of handy tasks that can be used to perform various operations on Algorand blockchain.

## Features

- [Wallet Aliasing](./tasks/wallet.md) - Manage your Algorand addresses and accounts effortlessly with the AlgoKit Wallet feature. This feature allows you to create short aliases for your addresses and accounts on AlgoKit CLI.
- [Vanity Address Generation](./tasks/vanity_address.md) - Generate vanity addresses for your Algorand accounts with the AlgoKit Vanity feature. This feature allows you to generate Algorand addresses which contains a specific keyword of your choice.
- [Transfer Assets or Algos](./tasks/transfer.md) - Transfer Algos or Assets from one account to another with the AlgoKit Transfer feature. This feature allows you to transfer Algos or Assets from one account to another on Algorand blockchain.
- [Opt-(in|out) Assets](./tasks/opt.md) - Opt-in or opt-out of Algorand Asset(s). Supports single or multiple assets.
- [Signing transactions](./tasks/sign.md) - Sign goal clerk compatible Algorand transactions.
- [Sending transactions](./tasks/send.md) - Send signed goal clerk compatible Algorand transactions.
- [NFD lookups](./tasks/nfd.md) - Perform a lookup via NFD domain or address, returning the associated address or domain respectively using the AlgoKit CLI.
- [IPFS uploads](./tasks/ipfs.md) - Upload files to IPFS.
- [Asset minting](./tasks/mint.md) - Mint new fungible or non-fungible assets on Algorand.
- [Analyze TEAL code](./tasks/analyze.md) - Analyze TEAL code using [`tealer`](https://github.com/crytic/tealer) integration for common vulnerabilities.
