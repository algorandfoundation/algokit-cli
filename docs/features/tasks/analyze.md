# AlgoKit Task Analyze

The `analyze` task is a command-line utility that analyzes TEAL programs for common vulnerabilities using [Tealer](https://github.com/crytic/tealer) integration. It allows you to detect a range of common vulnerabilities in code written in TEAL. For full list of vulnerability detectors refer to [Tealer documentation](https://github.com/crytic/tealer?tab=readme-ov-file#detectors).

## Usage

```bash
algokit task analyze INPUT_PATHS [OPTIONS]
```

### Arguments

- `INPUT_PATHS`: Paths to the TEAL files or directories containing TEAL files to be analyzed. This argument is required.

### Options

- `-r, --recursive`: Recursively search for all TEAL files within any provided directories.
- `--force`: Force verification without the disclaimer confirmation prompt.
- `--diff`: Exit with a non-zero code if differences are found between current and last reports.
- `-o, --output OUTPUT_PATH`: Directory path where to store the reports of the static analysis.
- `-e, --exclude DETECTORS`: Exclude specific vulnerabilities from the analysis. Supports multiple exclusions in a single run.

## Example

```bash
algokit task analyze ./contracts -r --exclude rekey-to --exclude missing-fee-check
```

This command will recursively analyze all TEAL files in the `contracts` directory and exclude the `missing-fee-check` vulnerability from the analysis.

## Security considerations

This task uses [`tealer`](https://github.com/crytic/tealer), a third-party tool, to suggest improvements for your TEAL programs, but remember to always test your smart contracts code, follow modern software engineering practices and use the [guidelines for smart contract development](https://developer.algorand.org/docs/get-details/dapps/smart-contracts/guidelines/). This should not be used as a substitute for an actual audit.
