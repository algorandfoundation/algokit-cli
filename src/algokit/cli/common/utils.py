# Cli/Click related helper functions and classes


import typing as t

import click

from algokit.cli.common.constants import ExplorerEntityType


class MutuallyExclusiveOption(click.Option):
    """
    A Click option that defines mutually exclusive command line options.

    Args:
        *args: Positional arguments passed to the parent class constructor.
        **kwargs: Keyword arguments passed to the parent class constructor.
            not_required_if (list): A list of options that the current option is mutually exclusive with.

    Attributes:
        not_required_if (list): A list of options that the current option is mutually exclusive with.

    Raises:
        AssertionError: If the `not_required_if` parameter is not provided.

    Example:
        ```python
        @click.command()
        @click.option('--option1', help='Option 1')
        @click.option('--option2', help='Option 2')
        @click.option('--option3', help='Option 3', cls=MutuallyExclusiveOption, not_required_if=['option1', 'option2'])
        def my_command(option1, option2, option3):
            # Command logic here
            pass
        ```

        In the example above, the `MutuallyExclusiveOption` class is used to define the `option3` command line option.
        This option is mutually exclusive with `option1` and `option2`,
        meaning that if either `option1` or `option2` is provided, `option3` cannot be used.
        If `option3` is provided along with `option1` or `option2`, a `click.UsageError` is raised.
    """

    def __init__(self, *args: t.Any, **kwargs: t.Any) -> None:
        self.not_required_if: list = kwargs.pop("not_required_if")

        assert self.not_required_if, "'not_required_if' parameter required"
        kwargs["help"] = (
            kwargs.get("help", "") + " Option is mutually exclusive with " + ", ".join(self.not_required_if) + "."
        ).strip()
        super().__init__(*args, **kwargs)

    def handle_parse_result(
        self, ctx: click.Context, opts: t.Mapping[str, t.Any], args: list[str]
    ) -> tuple[t.Any, list[str]]:
        """
        Overrides the `handle_parse_result` method of the `click.Option` class.

        This method checks if the current option is present in the provided options (`opts`)
        and if any of the mutually exclusive options are also present.
        If both the current option and a mutually exclusive option are present, it raises a `click.UsageError`.
        Otherwise, it returns the result of the parent `handle_parse_result` method.

        Args:
            ctx (click.Context): The Click context object.
            opts (dict): The dictionary of parsed options.
            args (list): The list of remaining arguments.

        Returns:
            The result of the parent `handle_parse_result` method.

        Raises:
            click.UsageError: If the current option and a mutually exclusive option are both present.
        """

        current_opt: bool = self.name in opts
        for mutex_opt in self.not_required_if:
            if mutex_opt in opts:
                if current_opt:
                    raise click.UsageError(
                        "Illegal usage: '" + str(self.name) + "' is mutually exclusive with " + str(mutex_opt) + "."
                    )
                self.prompt = None
        return super().handle_parse_result(ctx, opts, args)


def get_explorer_url(identifier: str | int, network: str, entity_type: ExplorerEntityType) -> str:
    """
    Returns a URL for exploring a specified type (transaction, asset, address) on the specified network.

    Args:
        identifier (str | int): The ID of the transaction, asset, or address to explore.
        network (str): The name of the network (e.g., "localnet", "testnet", "mainnet").
        entity_type (ExplorerEntityType): The type to explore (e.g., ExplorerEntityType.TRANSACTION,
        ExplorerEntityType.ASSET, ExplorerEntityType.ADDRESS).

    Returns:
        str: The URL for exploring the specified type on the specified network.

    Raises:
        ValueError: If the network or explorer type is invalid.
    """

    base_urls: dict[str, dict[str, str]] = {
        "testnet": {
            ExplorerEntityType.TRANSACTION.value: "https://testnet.algoexplorer.io/tx/",
            ExplorerEntityType.ASSET.value: "https://testnet.explorer.perawallet.app/assets/",
            ExplorerEntityType.ADDRESS.value: "https://testnet.algoexplorer.io/address/",
        },
        "mainnet": {
            ExplorerEntityType.TRANSACTION.value: "https://algoexplorer.io/tx/",
            ExplorerEntityType.ASSET.value: "https://explorer.perawallet.app/assets/",
            ExplorerEntityType.ADDRESS.value: "https://algoexplorer.io/address/",
        },
    }

    if network == "localnet":
        return f"https://app.dappflow.org/setnetwork?name=sandbox&redirect=explorer/{entity_type.value}/{identifier}/"

    if network not in base_urls:
        raise ValueError(f"Invalid network: {network}")

    if entity_type.value not in base_urls[network]:
        raise ValueError(f"Invalid explorer type: {entity_type}")

    return base_urls[network][entity_type.value] + str(identifier)
