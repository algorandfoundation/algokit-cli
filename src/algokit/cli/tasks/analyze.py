import json
import logging
import re
from pathlib import Path

import click

from algokit.core.tasks.analyze import (
    TEALER_SNAPSHOTS_ROOT,
    generate_report_filename,
    generate_summaries,
    generate_tealer_command,
    has_baseline_diff,
    install_tealer_if_needed,
    load_tealer_report,
    prepare_artifacts_folders,
    run_tealer,
)
from algokit.core.utils import run_with_animation

logger = logging.getLogger(__name__)


def display_analysis_summary(analysis_results: dict) -> None:
    """
    Display the summary of the analysis results.

    Args:
        analysis_results (dict): Dictionary containing analysis results.
    """
    impact_frequency: dict = {}
    for file_path, result_rows in analysis_results.items():
        click.echo(f"\nFile: {file_path}\n")
        for result in result_rows:
            click.echo(
                f"Detector: {result[0]}\n"
                f"Impact: {result[1]}\n"
                f"Details: {result[2]}\n"
                f"Execution Paths (#Lines):\n{result[3]}\n"
            )
            impact_frequency[result[1]] = impact_frequency.get(result[1], 0) + 1
    # print summary by impact label
    click.echo("\nTotal issues:")
    for impact, frequency in impact_frequency.items():
        click.secho(f"{impact}: {frequency}", fg="yellow")


def has_template_vars(path: Path) -> bool:
    """
    Check if the file contains template variables.

    Args:
        path (Path): The file path to check.

    Returns:
        bool: True if template variables are found, False otherwise.
    """
    content = path.read_text()
    return bool(re.search(r"^(?!.*//.*TMPL_).*TMPL_.*", content, flags=re.MULTILINE))


def get_input_files(*, input_paths: tuple[Path], recursive: bool) -> list[Path]:
    """
    Get input files based on the input paths and recursive flag.

    Args:
        input_paths (tuple[Path]): Tuple of input paths.
        recursive (bool): Flag to indicate recursive search.

    Returns:
        list[Path]: List of input files.
    """

    input_files = []
    for input_path in input_paths:
        if input_path.is_dir():
            pattern = "**/*.teal" if recursive else "*.teal"
            input_files.extend(sorted(input_path.glob(pattern)))
        else:
            if recursive:
                click.secho(
                    f"Warning: Ignoring recursive flag for {input_path} as it is not a directory.\n",
                    fg="yellow",
                )
            input_files.append(input_path)
    return sorted(set(input_files))


@click.command(
    name="analyze",
    help=(
        "Analyze TEAL programs for common vulnerabilities using Tealer. "
        "This task uses a third party tool to suggest improvements for your TEAL programs, "
        "but remember to always test your smart contracts code, follow modern software engineering practices "
        "and use the guidelines for smart contract development. "
        "This should not be used as a substitute for an actual audit. "
        "For full list of available detectors, please refer to https://github.com/crytic/tealer?tab=readme-ov-file#detectors"
    ),
)
@click.argument(
    "input_paths",
    nargs=-1,
    type=click.Path(exists=True, dir_okay=True, file_okay=True, path_type=Path),
    required=True,
)
@click.option(
    "-r",
    "--recursive",
    is_flag=True,
    help="Recursively search for all TEAL files within the provided directory.",
)
@click.option(
    "--force",
    is_flag=True,
    help="Force verification without the disclaimer confirmation prompt.",
)
@click.option(
    "--diff",
    "diff_only",
    is_flag=True,
    help=(
        "Exit with a non-zero code if differences are found between current "
        "and last reports. Reports are generated each run, but with this flag "
        "execution fails if the current report doesn't match "
        "the last report. Reports are stored in the "
        ".algokit/static-analysis/snapshots folder by default. Use --output for a "
        "custom path."
    ),
)
@click.option(
    "-o",
    "--output",
    "output_path",
    required=False,
    default=None,
    type=click.Path(dir_okay=True, file_okay=False, resolve_path=True, path_type=Path),
    help=(
        "Directory path where to store the results of the static analysis. "
        "Defaults to .algokit/static-analysis/snapshots."
    ),
)
@click.option(
    "-e",
    "--exclude",
    "detectors_to_exclude",
    multiple=True,
    default=[],
    type=click.STRING,
    help="Exclude specific vulnerabilities from the analysis. Supports multiple exclusions in a single run.",
)
def analyze(  # noqa: PLR0913, C901
    *,
    input_paths: tuple[Path],
    recursive: bool,
    force: bool,
    diff_only: bool,
    output_path: Path | None,
    detectors_to_exclude: list[str],
) -> None:
    """
    Analyze TEAL programs for common vulnerabilities using Tealer.
    """

    # Install tealer if needed
    install_tealer_if_needed()

    detectors_to_exclude = sorted(set(detectors_to_exclude))
    input_files = get_input_files(input_paths=input_paths, recursive=recursive)

    if not force:
        click.confirm(
            click.style(
                "Warning: This task uses `tealer` to suggest improvements for your TEAL programs, "
                "but remember to always test your smart contracts code, follow modern software engineering practices "
                "and use the guidelines for smart contract development. "
                "This should not be used as a substitute for an actual audit. Do you understand?",
                fg="yellow",
            ),
            default=True,
            abort=True,
        )

    reports = {}
    duplicate_files: dict[str, int] = {}
    prepare_artifacts_folders(output_path)
    total_files = len(input_files)
    for index in range(total_files):
        cur_file = input_files[index]
        file = cur_file.resolve()

        if has_template_vars(file):
            click.secho(
                f"Warning: Skipping {file} due to template variables. Substitute them before scanning.",
                err=True,
                fg="yellow",
            )
            continue

        filename = generate_report_filename(file, duplicate_files)

        # If a custom output path is provided, store the report in the specified path
        report_output_root = output_path or TEALER_SNAPSHOTS_ROOT
        report_output_path = report_output_root / filename

        command = generate_tealer_command(cur_file, report_output_path, detectors_to_exclude)
        old_report = load_tealer_report(str(report_output_path)) if report_output_path.exists() and diff_only else None
        if not old_report and diff_only:
            click.secho(
                f"Unable to provide the diff since {file} report is missing. "
                "Please run the task without the --diff flag first.",
                err=True,
                fg="red",
            )
            raise click.exceptions.Exit(1)

        try:
            run_with_animation(run_tealer, f"Analyzing {index + 1} out of {total_files} files", command)

            if diff_only and old_report:
                has_diff = has_baseline_diff(
                    cur_file=cur_file, report_output_path=report_output_path, old_report=old_report
                )
                if has_diff:
                    raise click.exceptions.Exit(1)

            reports[str(report_output_path.absolute())] = json.load(report_output_path.open())
        except Exception as e:
            if diff_only and old_report:
                report_output_path.write_text(json.dumps(old_report.model_dump(by_alias=True), indent=2))

            if isinstance(e, click.exceptions.Exit):
                raise e

            click.secho(
                f"An error occurred while analyzing {cur_file}. "
                "Please make sure the files supplied are valid TEAL code before trying again.",
                err=True,
                fg="red",
            )
            raise click.Abort("Error while running tealer") from e

    summaries = generate_summaries(reports, detectors_to_exclude=detectors_to_exclude)

    if summaries and not diff_only:
        display_analysis_summary(summaries)
        click.echo(f"Finished analyzing {total_files} files.")
        raise click.exceptions.Exit(1)
