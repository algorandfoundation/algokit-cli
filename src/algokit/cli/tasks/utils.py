# AlgoKit Tasks related utility functions

import logging
import os
import stat
import sys

import algosdk
import algosdk.encoding
import click
from algokit_utils import (
    Account,
    get_algod_client,
    get_algonode_config,
    get_default_localnet_config,
)
from algosdk.util import algos_to_microalgos, microalgos_to_algos

from algokit.cli.common.constants import AlgorandNetwork
from algokit.core.tasks.wallet import get_alias

logger = logging.getLogger(__name__)


def _validate_asset_balance(account_info: dict, asset_id: int, decimals: int, amount: int = 0) -> None:
    asset_record = next((asset for asset in account_info.get("assets", []) if asset["asset-id"] == asset_id), None)

    if not asset_record:
        raise click.ClickException("Account is not opted into the asset")

    if amount > 0 and asset_record["amount"] < amount:
        required = amount / 10**decimals
        available = asset_record["amount"] / 10**decimals
        raise click.ClickException(
            f"Insufficient asset balance in account, required: {required}, available: {available}"
        )


def _validate_algo_balance(account_info: dict, amount: int) -> None:
    if account_info.get("amount", 0) < amount:
        required = microalgos_to_algos(amount)  # type: ignore[no-untyped-call]
        available = microalgos_to_algos(account_info.get("amount", 0))  # type: ignore[no-untyped-call]
        raise click.ClickException(
            f"Insufficient Algos balance in account, required: {required} Algos, available: {available} Algos"
        )


def get_private_key_from_mnemonic() -> str:
    """
    Converts a mnemonic phrase into a private key.

    Returns:
        str: The private key generated from the mnemonic phrase.

    Raises:
        click.ClickException: If the entered mnemonic phrase is invalid.

    Example Usage:
        private_key = get_private_key_from_mnemonic()
        print(private_key)

    Inputs:
        None

    Flow:
        1. Prompts the user to enter a mnemonic phrase.
        2. Converts the entered mnemonic phrase into a private key using `algosdk.mnemonic.to_private_key`.
        3. Returns the private key.

    """

    mnemonic_phrase = click.prompt("Enter the mnemonic phrase (25 words separated by whitespace)", hide_input=True)
    try:
        return str(algosdk.mnemonic.to_private_key(mnemonic_phrase))  # type: ignore[no-untyped-call]
    except Exception as err:
        raise click.ClickException("Invalid mnemonic. Please provide a valid Algorand mnemonic.") from err


def load_algod_client(network: AlgorandNetwork) -> algosdk.v2client.algod.AlgodClient:
    """
    Returns an instance of the `algosdk.v2client.algod.AlgodClient` class for the specified network.

    Args:
        network (str): The network for which the `AlgodClient` instance needs to be loaded.

    Returns:
        algosdk.v2client.algod.AlgodClient: An instance of the `AlgodClient` class for the specified network.

    Raises:
        click.ClickException: If the specified network is invalid.
    """

    config_mapping = {
        AlgorandNetwork.LOCALNET: get_default_localnet_config("algod"),
        AlgorandNetwork.TESTNET: get_algonode_config("testnet", "algod", ""),
        AlgorandNetwork.MAINNET: get_algonode_config("mainnet", "algod", ""),
    }
    try:
        return get_algod_client(config_mapping[network])
    except KeyError as err:
        raise click.ClickException("Invalid network") from err


def get_asset_decimals(asset_id: int, algod_client: algosdk.v2client.algod.AlgodClient) -> int:
    """
    Retrieves the number of decimal places for a given asset.

    Args:
        asset_id (int): The ID of the asset for which the number of decimal places is to be retrieved. (0 for Algo)
        algod_client (algosdk.v2client.algod.AlgodClient): An instance of the AlgodClient class.

    Returns:
        int: The number of decimal places for the given asset.

    Raises:
        click.ClickException: If the asset info response is invalid.

    Example:
        asset_id = 123
        algod_client = algosdk.v2client.algod.AlgodClient("https://api.algoexplorer.io", "API_KEY")
        decimals = get_asset_decimals(asset_id, algod_client)
        print(decimals)
    """

    if asset_id == 0:
        return 6

    asset_info = algod_client.asset_info(asset_id)

    if not isinstance(asset_info, dict) or "params" not in asset_info or "decimals" not in asset_info["params"]:
        raise click.ClickException("Invalid asset info response")

    return int(asset_info["params"]["decimals"])


def validate_balance(
    algod_client: algosdk.v2client.algod.AlgodClient, account: Account | str, asset_id: int, amount: int = 0
) -> None:
    """
    Validates the balance of an account before an operation.

    Args:
        algod_client (algosdk.v2client.algod.AlgodClient): The AlgodClient object for
        interacting with the Algorand blockchain.
        account (Account | str): The account object.
        asset_id (int): The ID of the asset to be checked (0 for Algos).
        amount (int): The amount of Algos or asset for the operation. Defaults to 0 implying opt-in check only.

    Raises:
        click.ClickException: If any validation check fails.
    """
    address = account.address if isinstance(account, Account) else account
    account_info = algod_client.account_info(address)

    if not isinstance(account_info, dict):
        raise click.ClickException("Invalid account info response")

    if asset_id == 0:
        _validate_algo_balance(account_info, amount)
    else:
        decimals = get_asset_decimals(asset_id, algod_client)
        _validate_asset_balance(account_info, asset_id, decimals, amount)


def validate_address(address: str) -> None:
    """
    Check if a given address is a valid Algorand account address.

    Args:
        address (str): The address to be validated.

    Raises:
        click.ClickException: If the address is invalid.

    Returns:
        None
    """

    if not algosdk.encoding.is_valid_address(address):  # type: ignore[no-untyped-call]
        raise click.ClickException(f"`{address}` is an invalid account address")


def get_account_with_private_key(address: str) -> Account:
    """
    Retrieves an account object with the private key based on the provided address.

    Args:
        address (str): The address for which to retrieve the account object.

    Returns:
        Account: An account object with the address and private key.

    Raises:
        click.ClickException: If the address is not valid or if the alias does not exist or does not have a private key.
    """

    parsed_address = address.strip('"')

    try:
        validate_address(parsed_address)
        pk = get_private_key_from_mnemonic()
        return Account(address=parsed_address, private_key=pk)
    except click.ClickException as ex:
        alias_data = get_alias(parsed_address)

        if not alias_data:
            raise click.ClickException(f"Alias `{parsed_address}` alias does not exist.") from ex
        if not alias_data.private_key:
            raise click.ClickException(f"Alias `{parsed_address}` does not have a private key.") from ex

        return Account(address=alias_data.address, private_key=alias_data.private_key)


def get_address(address: str) -> str:
    """
    Validates the given address and returns it if valid. If the address is not valid,
    it checks if the address is an alias and returns the corresponding address if the alias exists.

    Args:
        address (str): The address to be validated or checked for an alias.

    Returns:
        str: The validated address or the corresponding address from an alias.

    Raises:
        click.ClickException: If the address is not valid and no corresponding alias exists.

    Example:
        address = get_address("ABCD1234")
        print(address)
    """

    parsed_address = address.strip('"')

    try:
        validate_address(parsed_address)
        return parsed_address
    except click.ClickException as ex:
        if len(parsed_address) == algosdk.constants.address_len:
            raise click.ClickException(f"`{parsed_address}` is an invalid account address") from ex

        alias_data = get_alias(parsed_address)

        if not alias_data:
            raise click.ClickException(f"Alias `{parsed_address}` alias does not exist.") from ex

        return alias_data.address


def stdin_has_content() -> bool:
    """
    Checks if there is content in the standard input.

    Returns:
        bool: True if there is content in the standard input, False otherwise.
    """

    mode = os.fstat(sys.stdin.fileno()).st_mode
    return stat.S_ISFIFO(mode) or stat.S_ISREG(mode)


def validate_account_balance_to_opt_in(
    algod_client: algosdk.v2client.algod.AlgodClient, account: Account, num_assets: int
) -> None:
    """
    Validates the balance of an account before opt in operation.
    Each asset requires 0.1 Algos to opt in.

    Args:
        algod_client (algosdk.v2client.algod.AlgodClient): The AlgodClient object for
        interacting with the Algorand blockchain.
        account (Account | str): The account object.
        num_assets (int): The number of the assets for opt in (0 for Algos).

    Raises:
        click.ClickException: If there is an insufficient fund in the account or account is not valid.
    """

    address = account.address if isinstance(account, Account) else account
    account_info = algod_client.account_info(address)

    if not isinstance(account_info, dict):
        raise click.ClickException("Invalid account info response")

    required_microalgos = num_assets * algos_to_microalgos(0.1)  # type: ignore[no-untyped-call]
    available_microalgos = account_info.get("amount", 0)
    if available_microalgos < required_microalgos:
        required_algo = microalgos_to_algos(required_microalgos)  # type: ignore[no-untyped-call]
        available_algos = microalgos_to_algos(available_microalgos)  # type: ignore[no-untyped-call]
        raise click.ClickException(
            f"Insufficient Algos balance in account to opt in, required: {required_algo} Algos, available:"
            f" {available_algos} Algos"
        )


def get_account_info(algod_client: algosdk.v2client.algod.AlgodClient, account_address: str) -> dict:
    account_info = algod_client.account_info(account_address)
    assert isinstance(account_info, dict)
    return account_info
