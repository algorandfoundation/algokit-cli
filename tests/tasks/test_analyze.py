import re
from collections.abc import Generator
from pathlib import Path
from unittest.mock import MagicMock, patch

import pytest
from algokit.cli.tasks.analyze import has_template_vars
from pytest_mock import MockerFixture

from tests.tasks.conftest import DUMMY_TEAL_FILE_CONTENT
from tests.utils.approvals import verify
from tests.utils.click_invoker import invoke
from tests.utils.proc_mock import ProcMock


def _format_snapshot(output: str, targets: list[str], replacement: str = "dummy") -> str:
    from algokit.core.utils import get_base_python_path

    python_base_path = get_base_python_path()
    if python_base_path is None:
        pytest.fail("Python base detection failed, this should work (even in CI)")

    output = output.replace(python_base_path, "python_base_path")

    for target in targets:
        output = output.replace(target, replacement)

    # If output contains more than one new line trim them to have at most one whitespace in between

    output = re.sub(r"^(pipx:|DEBUG: pipx:).*", "", output, flags=re.MULTILINE)
    return re.sub(r"\n\s*\n", "\n\n", output)


def _normalize_path(path: Path) -> str:
    return str(path.absolute()).replace("\\", r"\\")


@pytest.fixture(autouse=True)
def _disable_animation(mocker: MockerFixture) -> None:
    mocker.patch("algokit.core.utils.animate", return_value=None)


@pytest.fixture(autouse=True)
def cwd(tmp_path_factory: pytest.TempPathFactory) -> Generator[Path, None, None]:
    cwd = tmp_path_factory.mktemp("cwd", numbered=True)

    with patch("algokit.core.tasks.analyze.TEALER_REPORTS_ROOT", return_value=cwd), patch(
        "algokit.core.tasks.analyze.TEALER_SNAPSHOTS_ROOT", return_value=cwd
    ), patch("algokit.core.tasks.analyze.TEALER_DOT_FILES_ROOT", return_value=cwd):
        yield cwd


@pytest.fixture()
def generate_report_filename_mock() -> Generator[MagicMock, None, None]:
    with patch("algokit.cli.tasks.analyze.generate_report_filename", return_value="dummy_report.json") as mock:
        yield mock


@pytest.mark.usefixtures("generate_report_filename_mock")
def test_analyze_single_file(
    cwd: Path,
) -> None:
    teal_file = cwd / "dummy.teal"
    teal_file.write_text(DUMMY_TEAL_FILE_CONTENT)
    result = invoke(f"task analyze {_normalize_path(teal_file)} --output {_normalize_path(cwd)}", input="y\n", cwd=cwd)

    assert result.exit_code == 1
    result.output = _format_snapshot(
        result.output,
        [
            str(cwd),
        ],
    )
    verify(result.output)


def test_analyze_multiple_files(
    cwd: Path,
    generate_report_filename_mock: MagicMock,
) -> None:
    generate_report_filename_mock.side_effect = [f"dummy_{i}.teal" for i in range(5)]
    teal_folder = cwd / "dummy_contracts"
    teal_folder.mkdir()
    for i in range(5):
        teal_file = teal_folder / f"dummy_{i}.teal"
        teal_file.write_text(DUMMY_TEAL_FILE_CONTENT)
    result = invoke(
        f"task analyze {_normalize_path(teal_folder)} --output {_normalize_path(cwd)}", input="y\n", cwd=cwd
    )

    assert result.exit_code == 1
    for i in range(5):
        result.output = result.output.replace(str(teal_folder / f"dummy_{i}.teal"), f"dummy_contracts/dummy_{i}.teal")
    result.output = _format_snapshot(
        result.output,
        [
            str(cwd),
        ],
    )
    verify(result.output)


def test_analyze_multiple_files_recursive(
    cwd: Path,
    generate_report_filename_mock: MagicMock,
) -> None:
    teal_root_folder = cwd / "dummy_contracts"
    generate_report_filename_mock.side_effect = [teal_root_folder / f"subfolder_{i}/dummy.teal" for i in range(5)]

    for i in range(5):
        teal_folder = teal_root_folder / f"subfolder_{i}"
        teal_folder.mkdir(parents=True)
        teal_file = teal_folder / "dummy.teal"
        teal_file.write_text(DUMMY_TEAL_FILE_CONTENT)
    result = invoke(
        f"task analyze {_normalize_path(teal_root_folder)} --recursive --output {_normalize_path(cwd)}",
        input="y\n",
        cwd=cwd,
    )

    assert result.exit_code == 1
    for i in range(5):
        result.output = re.sub(r"^File: .*", f"File:  {i}_dummy.teal", result.output, flags=re.MULTILINE)
    result.output = _format_snapshot(
        result.output,
        [str(cwd)],
        "dummy_file.teal",
    )
    verify(result.output)


@pytest.mark.usefixtures("generate_report_filename_mock")
def test_exclude_vulnerabilities(
    cwd: Path,
) -> None:
    teal_file = cwd / "dummy.teal"
    teal_file.write_text(DUMMY_TEAL_FILE_CONTENT)
    result = invoke(
        f"task analyze {_normalize_path(teal_file)} --exclude is-deletable "
        f"--exclude rekey-to --exclude missing-fee-check --output {_normalize_path(cwd)}",
        input="y\n",
        cwd=cwd,
    )

    assert result.exit_code == 0
    result.output = _format_snapshot(result.output, [str(cwd)])
    verify(result.output)


def test_analyze_skipping_tmpl_vars(
    cwd: Path,
) -> None:
    teal_file = cwd / "dummy.teal"
    teal_file.write_text(
        DUMMY_TEAL_FILE_CONTENT.replace("pushint 4 // UpdateApplication", "pushint TMPL_VAR // UpdateApplication")
    )
    result = invoke(f"task analyze {_normalize_path(teal_file)}", input="y\n", cwd=cwd)

    assert result.exit_code == 0
    result.output = _format_snapshot(result.output, [str(cwd)])
    verify(result.output)


def test_analyze_commented_tmpl_vars(
    cwd: Path,
) -> None:
    teal_file = cwd / "dummy.teal"
    teal_file.write_text(
        DUMMY_TEAL_FILE_CONTENT.replace("pushint 4 // UpdateApplication", "pushint TMPL_VAR // UpdateApplication")
    )
    assert has_template_vars(teal_file)

    teal_file.write_text(DUMMY_TEAL_FILE_CONTENT.replace("pushint 4 // UpdateApplication", "pushint 4 // TMPL_VAR"))
    assert not has_template_vars(teal_file)


def test_analyze_abort_disclaimer(
    cwd: Path,
) -> None:
    teal_file = cwd / "dummy.teal"
    teal_file.touch()
    result = invoke(f"task analyze {_normalize_path(teal_file)} --output {_normalize_path(cwd)}", input="n\n", cwd=cwd)

    assert result.exit_code == 1
    verify(result.output)


def test_analyze_error_in_tealer(
    cwd: Path,
    mocker: MockerFixture,
) -> None:
    mocker.patch("algokit.cli.tasks.analyze.run_tealer", side_effect=Exception("dummy"))
    teal_file = cwd / "dummy.teal"
    teal_file.touch()
    result = invoke(f"task analyze {_normalize_path(teal_file)} --output {_normalize_path(cwd)}", input="y\n", cwd=cwd)

    assert result.exit_code == 1
    result.output = _format_snapshot(result.output, [str(cwd)])
    verify(result.output)


@pytest.mark.usefixtures("generate_report_filename_mock")
def test_analyze_diff_flag(
    cwd: Path,
) -> None:
    teal_file = cwd / "dummy.teal"
    teal_file.write_text(DUMMY_TEAL_FILE_CONTENT)
    result = invoke(f"task analyze {_normalize_path(teal_file)} --output {_normalize_path(cwd)}", input="y\n", cwd=cwd)
    assert result.exit_code == 1

    teal_file.write_text("\n#pragma version 8\nint 1\nreturn\n")
    result = invoke(
        f"task analyze {_normalize_path(teal_file)} --diff --output {_normalize_path(cwd)}", input="y\n", cwd=cwd
    )
    assert result.exit_code == 1
    result.output = _format_snapshot(result.output, [str(cwd)])
    verify(result.output)


@pytest.mark.usefixtures("generate_report_filename_mock")
def test_analyze_diff_flag_missing_old_report(
    cwd: Path,
) -> None:
    teal_file = cwd / "dummy.teal"
    teal_file.write_text(DUMMY_TEAL_FILE_CONTENT)
    result = invoke(
        f"task analyze {_normalize_path(teal_file)} --diff --output {_normalize_path(cwd)}", input="y\n", cwd=cwd
    )
    assert result.exit_code == 1
    result.output = _format_snapshot(result.output, [str(cwd)])
    verify(result.output)


def test_analyze_error_no_pipx(cwd: Path, mocker: MockerFixture, proc_mock: ProcMock) -> None:
    proc_mock.should_fail_on("tealer --version")
    mocker.patch("algokit.core.utils.get_candidate_pipx_commands", return_value=[])

    teal_file = cwd / "dummy.teal"
    teal_file.touch()
    result = invoke(f"task analyze {_normalize_path(teal_file)}", input="y\n", cwd=cwd)

    assert result.exit_code == 1
    result.output = _format_snapshot(result.output, [str(cwd)])
    verify(result.output)
