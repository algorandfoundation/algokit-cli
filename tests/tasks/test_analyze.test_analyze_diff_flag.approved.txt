DEBUG: Running 'tealer --version' in '{current_working_directory}'
DEBUG: tealer: 0.1.2
Warning: This task uses `tealer` to suggest improvements for your TEAL programs, but remember to always test your smart contracts code, follow modern software engineering practices and use the guidelines for smart contract development. This should not be used as a substitute for an actual audit. Do you understand? [Y/n]: y
DEBUG: Running 'tealer --json {current_working_directory}/dummy_report.json detect --contracts {current_working_directory}/dummy.teal' in '{current_working_directory}'
DEBUG: tealer: Reading contract from file: "{current_working_directory}/dummy.teal"
DEBUG: tealer: json output is written to {current_working_directory}/dummy_report.json
ERROR: Diff detected in {current_working_directory}/dummy.teal! Please check the content of the snapshot report {current_working_directory}/dummy_report.json against the latest received report at {current_working_directory}/dummy_report.received.json.
